### PR TITLE
Add unpack-dfu command to stm32 port mboot tools

### PR DESCRIPTION
**Pull request by davidmm01**

### CODE REVIEW

**Comment by jej** [comment]
Test failed...

`stm32: Add unpack-dfu command to stm32 port mboot tools.`

:)

File: ports/stm32/mboot/mboot_pack_dfu.py [hunk 1]
```diff
@@ -228,11 +230,22 @@ def verify_pack_dfu(keys, filename):
...
+    return binary_data
+
+
+def unpack_dfu(keys, args):
+    # Load previously generated keys.
+    keys.load()
+
+    # build a DFU file from the decrypted binary data
+    binary_data = verify_pack_dfu(keys, args.infile[0])
+    dfu.build(args.outfile[0], [[{"address": 0x08008000, "data": binary_data}]])
```
Comment (dpgeorge) [inline ports/stm32/mboot/mboot_pack_dfu.py:248]: The 0x08008000 cannot be hard coded.  `verify_pack_dfu` must return this value.

File: ports/stm32/mboot/mboot_pack_dfu.py [hunk 2]
```diff
@@ -228,11 +230,22 @@ def verify_pack_dfu(keys, filename):
                 chunk, 0, MBOOT_PACK_HYDRO_CONTEXT, keys.secretbox
             )
             assert chunk is not None
+            binary_data += chunk
```
Comment (dpgeorge) [inline ports/stm32/mboot/mboot_pack_dfu.py:233]: This needs to go after the decompress below.

File: ports/stm32/mboot/mboot_pack_dfu.py [hunk 3]
```diff
@@ -228,11 +230,22 @@ def verify_pack_dfu(keys, filename):
...
             if header[1] == MBOOT_PACK_CHUNK_FW_GZIP:
                 chunk = zlib.decompress(chunk, wbits=-15)
             full_sig.update(chunk)
     full_sig_pass = full_sig.final_verify(actual_sig, keys.sign_pk)
     assert full_sig_pass
+    return binary_data
```
Comment (dpgeorge) [inline ports/stm32/mboot/mboot_pack_dfu.py:239]: This should return a list of tuples of addr, data, to feed into `dfu.build()`.  But I think it's ok for now to just support one element, ie just one contiguous chunk with one address.

That address would be the first address encountered in the for loop above, eg `base_addr = addr`.  And then to assert that there's only one element, subsequent iterations of the for loop would assert `addr == base_addr + len(binary_data)`.  Then at the end return `[{"address": base_addr, "data": binary_data}]`.

**Comment by dpgeorge** [comment]
Would be good to add a test to CI.  Maybe something like this:
```diff
diff --git a/tools/ci.sh b/tools/ci.sh
index 4f9955146..e258e12a0 100755
--- a/tools/ci.sh
+++ b/tools/ci.sh
@@ -223,6 +223,8 @@ function ci_stm32_nucleo_build {
     make ${MAKEOPTS} -C ports/stm32 BOARD=NUCLEO_L476RG DEBUG=1
     make ${MAKEOPTS} -C ports/stm32 BOARD=NUCLEO_WB55
     make ${MAKEOPTS} -C ports/stm32/mboot BOARD=NUCLEO_WB55
+    python3 ports/stm32/mboot/mboot_pack_dfu.py -k ports/stm32/NUCLEO_WB55/mboot_keys.h unpack-dfu ports/stm32/build-NUCLEO_WB55/firmware.pack.dfu ports/stm32/build-NUCLEO_WB55/firmware.unpack.dfu
+    diff ports/stm32/build-NUCLEO_WB55/firmware.unpack.dfu ports/stm32/build-NUCLEO_WB55/firmware.dfu
 }
 
 ########################################################################################
```

File: ports/stm32/mboot/mboot_pack_dfu.py (hunk 1 above)
Comment (davidmm01) [reply ports/stm32/mboot/mboot_pack_dfu.py]: Now returned from `verify_pack_dfu`

File: ports/stm32/mboot/mboot_pack_dfu.py (hunk 2 above)
Comment (davidmm01) [reply ports/stm32/mboot/mboot_pack_dfu.py]: Moved after decompress

File: ports/stm32/mboot/mboot_pack_dfu.py [hunk 4]
```diff
@@ -231,8 +239,21 @@ def verify_pack_dfu(keys, filename):
...
+    return [{"address": base_addr, "data": binary_data}]
+
+
+def unpack_dfu(keys, args):
+    # Load previously generated keys.
+    keys.load()
+
+    # build a DFU file from the decrypted binary data
+    data = verify_pack_dfu(keys, args.infile[0])
+    dfu.build(args.outfile[0], [[{"address": data[0]["address"], "data": data[0]["data"]}]])
```
Comment (dpgeorge) [inline ports/stm32/mboot/mboot_pack_dfu.py:256]: much simpler to do: `dfu.build(args.outfile[0], [data])`

File: ports/stm32/mboot/mboot_pack_dfu.py (hunk 3 above)
Comment (davidmm01) [reply ports/stm32/mboot/mboot_pack_dfu.py]: Done

File: ports/stm32/mboot/mboot_pack_dfu.py (hunk 4 above)
Comment (davidmm01) [reply ports/stm32/mboot/mboot_pack_dfu.py]: Thanks, done

**Comment by dpgeorge** [comment]
Neat addition, thanks!  Merged in 2eed9780ba7074de9e464a2bc771ad14f0332a6c

**Pull request closed without merging**